### ISSUE TITLE
Abbreviation starting with backslash isn't triggered

### ISSUE THREAD

**Issue by choco-latin** [autokey/autokey#735]
### Has this issue already been reported?

- [X] I have searched through the existing issues.

### Is this a question rather than an issue?

- [X] This is not a question.

### What type of issue is this?

Bug

### Which Linux distribution did you use?

Ubuntu 22.04 with session started with Xorg (French keyboard)

### Which AutoKey GUI did you use?

GTK

### Which AutoKey version did you use?

0.96.0

### How did you install AutoKey?

downloaded .deb from the repo and isntalled with dpkg as [recommended](https://www.autohotkey.com/boards/viewtopic.php?t=65552)

### Can you briefly describe the issue?

I defined a phrase whose abbreviation is `\RR`. After saving and testing, nothing happens. Instead, if the abbreviation is set to be simply `RR` it works exactly as intended. So I deduce that there is something special about `\`.

### Can the issue be reproduced?

Always

### What are the steps to reproduce the issue?

- create new phrase
- abbreviation is `\RR`
  - remove typed abbreviation
  - trigger immediately
  - trigger when typed as part of a phrase
- paste using Clipboard

### What should have happened?

AK should be triggered

### What actually happened?

AK wasn't triggered

### Do you have screenshots?

No

### Anything else?

- Is backslash a special character needing to be escaped? I searched in the wiki but couldn't find anything about it
- I am using a FR AZERTY keyboad, could it be an issue?

**Comment by josephj11** [comment]
TL;DR: Don't do that.

Not sure what's actually going on, but using a backslash in an abbreviation is a bad idea in general. Most programs in Linux treat the backslash as a metacharacter which signifies that the next character should be treated as a literal. So, for all practical purposes, `\R` and `R` are the same character. To get a literal backslash, you have to escape it too with something like `\\`.

E.G.: if you need to type a command on the terminal command line (CLI) that contains a hash mark, you have to escape it with a backslash ... or the hash mark will make the CLI treat the rest of the line as a comment.

What that does to AutoKey's trigger abbreviations, I don't know.

Your keyboard layout should be fine _if_ it's your primary (or only) keyboard layout. AutoKey tends to ignore secondary keyboard layouts.

Here's a wild guess at what might be happening: For some reason, AutoKey is treating the backslash as a literal in the abbreviation (which, to my mind, is a bug). But when you type that abbreviation into almost anything, it gets converted into ithe sames string without the backslash before AutoKey gets to see it, so it doesn't match anything.

If you really want to, you can verify what AutoKey did with the abbreviation by looking at your phrase in the AutoKey main window to see if the backslash is still in the abbreviation there.

You can also run an AutoKey [trace](https://github.com/autokey/autokey-python2/wiki/Problem-Reporting-Guide) to see what input AutoKey actually sees when you type your abbreviation.

**Comment by Elliria** [comment]
It works for me in **Kubuntu 22.04 LTS** with **AutoKey 0.95.10** and also with **AutoKey 0.96.0**, but I use a US keyboard rather than a French keyboard. The keyboard may be what's interfering here.

**Comment by choco-latin** [comment]
Hi,

@josephj11 here is a trace of Autokey when I type \RR : 

```
2026-04-19 20:16:16,204 DEBUG - autokey.iomediator.iomediator - Key.ALT_GR pressed
2026-04-19 20:16:16,503 DEBUG - autokey.service - Raw key: '_', modifiers: [<Key.ALT_GR: '<alt_gr>'>], Key: \
2026-04-19 20:16:16,503 DEBUG - autokey.service - Window visible title: '*Unsaved Document 1', Window class: 'xed.Xed'
2026-04-19 20:16:16,685 DEBUG - autokey.iomediator.iomediator - Key.ALT_GR released
2026-04-19 20:16:17,220 DEBUG - autokey.iomediator.iomediator - Key.SHIFT pressed
2026-04-19 20:16:17,587 DEBUG - autokey.service - Raw key: 'r', modifiers: [<Key.SHIFT: '<shift>'>], Key: R
2026-04-19 20:16:17,588 DEBUG - autokey.service - Window visible title: '*Unsaved Document 1', Window class: 'xed.Xed'
2026-04-19 20:16:17,588 DEBUG - autokey.service - Input queue at end of handle_keypress: deque(['R'], maxlen=150)
2026-04-19 20:16:17,740 DEBUG - autokey.service - Raw key: 'r', modifiers: [<Key.SHIFT: '<shift>'>], Key: R
2026-04-19 20:16:17,740 DEBUG - autokey.service - Window visible title: '*Unsaved Document 1', Window class: 'xed.Xed'
2026-04-19 20:16:17,740 DEBUG - autokey.service - Input queue at end of handle_keypress: deque(['R', 'R'], maxlen=150)
2026-04-19 20:16:17,927 DEBUG - autokey.iomediator.iomediator - Key.SHIFT released
```

For reference, on a french keyboard the key \ is accessible via alt_gr on the key containing also _ and 8, so that might explain the  Raw key '_' ?

I also note that there is no equivalent of `Input queue at end of handle_keypress: deque(['R', 'R'], maxlen=150)` for the backslash?

> Your keyboard layout should be fine if it's your primary (or only) keyboard layout. AutoKey tends to ignore secondary keyboard layouts.

yes I only use one keyboard (French)

@Elliria 

> It works for me in Kubuntu 22.04 LTS with AutoKey 0.95.10 and also with AutoKey 0.96.0, but I use a US keyboard rather than a French keyboard. The keyboard may be what's interfering here.

This comment made me want to try what happens when using a US layout . I defined a US layout (for which the backslash is acessible without altgr) and .. it worked : 

```
2026-04-19 20:38:11,534 DEBUG - autokey.service - Raw key: '\\', modifiers: [], Key: \
2026-04-19 20:38:11,534 DEBUG - autokey.service - Window visible title: '*Unsaved Document 1', Window class: 'xed.Xed'
2026-04-19 20:38:11,535 DEBUG - autokey.service - Input queue at end of handle_keypress: deque(['\\'], maxlen=150)
2026-04-19 20:38:12,310 DEBUG - autokey.iomediator.iomediator - Key.SHIFT pressed
2026-04-19 20:38:12,789 DEBUG - autokey.service - Raw key: 'r', modifiers: [<Key.SHIFT: '<shift>'>], Key: R
2026-04-19 20:38:12,790 DEBUG - autokey.service - Window visible title: '*Unsaved Document 1', Window class: 'xed.Xed'
2026-04-19 20:38:12,790 DEBUG - autokey.service - Input queue at end of handle_keypress: deque(['\\', 'R'], maxlen=150)
2026-04-19 20:38:13,117 DEBUG - autokey.service - Raw key: 'r', modifiers: [<Key.SHIFT: '<shift>'>], Key: R
2026-04-19 20:38:13,117 DEBUG - autokey.service - Window visible title: '*Unsaved Document 1', Window class: 'xed.Xed'
2026-04-19 20:38:13,117 INFO - autokey.service - Matched Phrase "mathbb" having abbreviations "['\\RR']" against current input
2026-04-19 20:38:13,118 DEBUG - autokey.iomediator.iomediator - Send via clipboard
2026-04-19 20:38:13,118 DEBUG - autokey.interface - Sending string via clipboard: Enter phrase contents
2026-04-19 20:38:13,118 DEBUG - autokey.interface - Sending via clipboard enqueued.
2026-04-19 20:38:13,119 DEBUG - autokey.service - Input queue at end of handle_keypress: deque([], maxlen=150)
2026-04-19 20:38:13,121 DEBUG - autokey.interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2026-04-19 20:38:13,123 DEBUG - autokey.interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2026-04-19 20:38:13,124 DEBUG - autokey.interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2026-04-19 20:38:13,130 DEBUG - autokey.iomediator.iomediator - Send via event interface
2026-04-19 20:38:13,132 DEBUG - autokey.interface - Send modified key: modifiers: ['<ctrl>'] key: v
2026-04-19 20:38:13,524 DEBUG - autokey.iomediator.iomediator - Key.SHIFT released
```

This being said, yet another funny thing happened when I wanted to send you a simpler trace : instead of doing shift before typing r, using MAJ in the first place will make things simpler. And there I got a failure:

```
2026-04-19 20:39:56,362 DEBUG - autokey.iomediator.iomediator - Key.CAPSLOCK pressed
2026-04-19 20:39:56,441 DEBUG - autokey.iomediator.iomediator - Key.CAPSLOCK released
2026-04-19 20:40:04,092 DEBUG - autokey.service - Raw key: '\\', modifiers: [], Key: |
2026-04-19 20:40:04,092 DEBUG - autokey.service - Window visible title: '*Unsaved Document 1', Window class: 'xed.Xed'
2026-04-19 20:40:04,092 DEBUG - autokey.service - Input queue at end of handle_keypress: deque(['|'], maxlen=150)
2026-04-19 20:40:05,762 DEBUG - autokey.service - Raw key: 'r', modifiers: [], Key: R
2026-04-19 20:40:05,762 DEBUG - autokey.service - Window visible title: '*Unsaved Document 1', Window class: 'xed.Xed'
2026-04-19 20:40:05,762 DEBUG - autokey.service - Input queue at end of handle_keypress: deque(['|', 'R'], maxlen=150)
2026-04-19 20:40:06,113 DEBUG - autokey.service - Raw key: 'r', modifiers: [], Key: R
2026-04-19 20:40:06,114 DEBUG - autokey.service - Window visible title: '*Unsaved Document 1', Window class: 'xed.Xed'
2026-04-19 20:40:06,114 DEBUG - autokey.service - Input queue at end of handle_keypress: deque(['|', 'R', 'R'], maxlen=150)
```

So in hindsight this was obviously a mistake : typing maj + backslash should produce a |. But on linux there is this weird behavior that CAPSLOCK doesn't behave like shift. With the US keyboard, when using capslock and pressing backslash, I do get a backslash, but from AutoKey's point of view that seem to make a difference. In summary:

- French keyboard, pressing altgr and '_' gives `Raw key: '_', modifiers: [<Key.ALT_GR: '<alt_gr>'>], Key: \` (not working)
- US keyboard pressing '\' gives `Raw key: '\\', modifiers: [], Key: \` (working as expected)
- US keyboard pressing CAPSLOCK and '\' gives `Raw key: '\\', modifiers: [], Key: |` (not working)

I hope all this makes sense to any of you! Oh, and about this

> TL;DR: Don't do that. [...] using a backslash in an abbreviation is a bad idea in general.

I am a scientist and using Latex , a language for which all macros and functions are defined starting with a backslash. It is common to want to use latex code outside of latex itself (mail, markdown etc) which made autohotkey an essential tool for my daily work. This is also without saying that in the current era of collaborative editing, it is much more convenient that everyone has its own autohotkey than one of the authors imposing their macro to the others.

**Comment by josephj11** [comment]
I haven't reviewed this entire thread, but here's what jumps out at me:
IDK AutoKey internals, but I believe it was written without a lot of consideration of non-EN-US charactersets as evidencced by a large number of bugs we've fixed in that area. In particular, I don't think an abbreviation can contain any multibyte characters like backslash with AltGR probably is (Running xev will show you). I'm surprised if that ever worked. Until a very recent fix, AltGR wasn't even recognized as a modifier key for hotkeys...

Also, since xedit starts with an x, I'm a bit surprised that AutoKey works at all with it because native x11 apps tend to ignore synthetic keypresses by default (those are what the Python keyboard module emits).

Our whole keyboard interface needs a rewrite, but some of the problems with it appear to be known problems deeper in Python that nobody wants to fix.

IDK how to address this issue. The only thing I can think of would be to remap the backslash key using xmodmap or xkb into something AutoKey can recognize, but that cure seems worse than the disease. My guess is that someone very clever could patch something into AutoKey that would make it handle this as an exception, but that it probably wouldn't be something we'd want to incorporate into the code in general.

**Comment by Elliria** [comment]
I'm glad that you got it working with the US layout, but is that a feasible solution for you? If not, then remapping some of the keys on your preferred layout may be the only way to do it.

By the way, your mention of the **Capslock** key not behaving the same way as the **Shift** key reminded me of [this section of the AutoKey wiki](https://github.com/autokey/autokey/wiki/Key-Combinations#effect-of-keyboardsend_key-and-keyboardsend_keys-on-case) that you might find interesting and that might help to unravel some of this.

The way it works is that AutoKey handles keys differently based on whether they're double-value or single-value keys in order to make those double values accessible and their shifted or capslocked status is used for that. The `\` key is a double-value key, so it's subject to being interpreted by AutoKey in that way.

Depending on how intricate the code for that behavior is, there might be a way to hack it to turn the behavior off, with, of course, a warning that the double values would no longer be available from the keyboard.

It might not be a bad idea to create a feature-request for a toggle that could turn that AutoKey behavior on and off. Then, you could use the toggle to intercept AutoKey's native behavior temporarily and restore it afterwards, with the only potential hitch being if you happened to want both behaviors at once.

**Comment by choco-latin** [comment]
> I'm glad that you got it working with the US layout, but is that a feasible solution for you? 

Not really unfrotunately

> If not, then remapping some of the keys on your preferred layout may be the only way to do it.

Thanks for the suggestion, but quoting @josephj11 the cure seems worse than the disease !

> It might not be a bad idea to create a feature-request for a toggle that could turn that AutoKey behavior on and off.

I have tried to understand your suggestion but I am going to be honest here, I barely understand what it is about. 

It makes sense to me that AutoKey wants to both know what key was pressed vs. what key was "obtained" as a result of modifiers. The two API `keyboard.send_key()` and `keyboard.send_keys()` fulfill that role. What I do not understand is why would basic macros not work when using double-value keys on a non-US keyboard. 

If AutoKey was "hard-coded" with only the US keyboard in mind as suggested by @josephj11 , maybe this should be properly disclaimed somewhere?

**Comment by josephj11** [comment]
IIUC, It's AutoKey's faullt that  non -EN-US characters don't work in abbreviations, but it's Python's fault that they don't work in the keyboard modules. The latter is relatively easy to work around. IDK what would be required to fix the former, which is the problem here.

**Comment by Elliria** [comment]
A **double-value** key can produce either one character or another character depending on whether you press the **Shift** key with it or not. For example, the **8** key can produce an asterisk if you press the **Shift** key while pressing the **8** key.

The AutoKey `keyboard.send_key()` and `keyboard.send_keys()` API calls either do or don't press the **Shift** key as part of their action depending on which of the two values you're sending in the API calls. For example, if you send an **8**, they don't press the **Shift** key and if you send an asterisk, they do press the **Shift** key.

Since you're working with a phrase rather than a script, this shouldn't apply, but it occurs to me that since AutoKey is aware of the dual values on those keys and since each of those keys can only have one of its two values at any given moment in time, it's possible that it's "choosing" one that conflicts with what you're asking it to do by having a double-value key in your abbreviation specifically because of the difference in your keyboard in comparison with our keyboards. In other words, the state of that key may be something else entirely than the backslash that you expect it to be. I'm basically just thinking out loud here and could be wildly off the mark.

Adding a disclaimer is a great idea. I added the [Effects of keyboard layouts on key behavior](https://github.com/autokey/autokey/wiki/Key-Combinations#effects-of-keyboard-layouts-on-key-behavior) section to the [Key Combinatons](https://github.com/autokey/autokey/wiki/Key-Combinations) wiki page. Feel free to add to it or improve it in case I missed something.
